components/feature/session/src/main/java/mozilla/components/feature/session/view/VerticalSwipeRefreshLayout.kt

```diff
@@ -0,0 +1,180 @@
+            }
+        }
+
+        return callSuperOnInterceptTouchEvent(event)
+    }
+
+    override fun onStartNestedScroll(child: View, target: View, nestedScrollAxes: Int): Boolean {
+        // Ignoring nested scrolls from descendants.
+        // Allowing descendants to trigger nested scrolls would defeat the purpose of this class
+        // and result in pull to refresh to happen for all movements on the Y axis
+        // (even as part of scale/quick scale gestures) while also doubling the throbber with the overscroll shadow.
+        return if (isEnabled) {
+            return false
+        } else {
+            callSuperOnStartNestedScroll(child, target, nestedScrollAxes)
+        }
+    }
+
+    @VisibleForTesting
+    internal fun maybeAddDoubleTapEvent(event: MotionEvent) {
+        val currentEventAction = event.action
+
+        // A double tap event must follow the order:
+        // ACTION_DOWN - ACTION_UP - ACTION_DOWN
+        // all these events happening in an interval defined by a system constant - DOUBLE_TAP_TIMEOUT
+
+        if (MotionEvent.ACTION_DOWN == currentEventAction) {
+            if (quickScaleEvents.second != null) {
+                if (event.eventTime - quickScaleEvents.second!!.eventTime > doubleTapTimeout) {
+                    // Too much time passed for the MotionEvents sequence to be considered
+                    // a quick scale gesture. Restart counting.
+                    forgetQuickScaleEvents()
+                    quickScaleEvents = quickScaleEvents.copy(first = MotionEvent.obtain(event))
+                } else {
+                    quickScaleEvents = quickScaleEvents.copy(third = MotionEvent.obtain(event))
+                }
+            } else {
+                // This may be the first time the user touches the screen or
+                // the gesture was not finished with ACTION_UP.
+                forgetQuickScaleEvents()
+                quickScaleEvents = quickScaleEvents.copy(first = MotionEvent.obtain(event))
+            }
+        }
+        // For the double tap events series we need ACTION_DOWN first
+        // and then ACTION_UP second.
+        else if (MotionEvent.ACTION_UP == currentEventAction && quickScaleEvents.first != null) {
+            quickScaleEvents = quickScaleEvents.copy(second = MotionEvent.obtain(event))
+        }
+    }
+
+    @VisibleForTesting
+    internal fun forgetQuickScaleEvents() {
+        quickScaleEvents.toList().forEach { it?.recycle() }
+        quickScaleEvents = Triple(null, null, null)
+        isQuickScaleInProgress = false
+    }
+
+    @VisibleForTesting
+    internal fun isQuickScaleInProgress(events: Triple<MotionEvent?, MotionEvent?, MotionEvent?>): Boolean {
+        return if (events.isNotNull()) {
+            isQuickScaleInProgress(events.first!!, events.second!!, events.third!!)
+        } else {
+            false
+        }
+    }
+
+    // Method closely following GestureDetectorCompat#isConsideredDoubleTap.
```

Comment:
It's preferable to use KDoc comments here.
```suggestion
    /**
     * Method closely following GestureDetectorCompat#isConsideredDoubleTap.
     * Allows for serial inference of double taps as opposed to using callbacks.
     */
    @VisibleForTesting
```

Reply:
I prefer to use kdoc/javadoc for public documentation and these type of comment for future maintainers of the code.
The above doesn't really serve as method documentation but rather as implementation specifics.